File: tests/08-execute_validations_checks.ts
import * as anchor from "@anchor-lang/core";
import { Program } from "@anchor-lang/core";
import { AccountMeta, Keypair, TransactionInstruction } from "@solana/web3.js";
import assert from "assert";

import { R3SessionKeys } from "../target/types/r3_session_keys";
import * as sessionKeys from "./utils/pda";
import {
  ensureProgramInitialized,
  executeWithSession,
  sendExpectError,
  timestampFromFuture,
} from "./utils/helpers";
import {
  ANCHOR_DISCRIMINATOR_LEN,
  DUMMY_ANCHOR_DISCRIMINATOR,
  TARGET_PROGRAM_PLACEHOLDER,
} from "./utils/constants";

describe("08 - Execute validation checks", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const baseWallet = provider.wallet as anchor.Wallet;
  const program = anchor.workspace.r3SessionKeys as Program<R3SessionKeys>;

  before(async () => {
    await ensureProgramInitialized(program, baseWallet);
  });

  async function createApprovedSession() {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();
    const userSmartWallet = sessionKeys.deriveUserSmartWalletPda(
      program.programId,
      smartWalletOwner.publicKey
    );

    await program.methods
      .createSmartWallet(smartWalletOwner.publicKey)
      .accounts({ admin: baseWallet.publicKey })
      .signers([baseWallet.payer])
      .rpc();

    await program.methods
      .createSession(
        sessionKey.publicKey,
        TARGET_PROGRAM_PLACEHOLDER,
        await timestampFromFuture(provider.connection),
        DUMMY_ANCHOR_DISCRIMINATOR,
        ANCHOR_DISCRIMINATOR_LEN
      )
      .accounts({
        sessionExecutor: baseWallet.publicKey,
        userSmartWallet,
      })
      .signers([baseWallet.payer])
      .rpc();

    await program.methods
      .approveSession(sessionKey.publicKey)
      .accounts({
        smartWalletOwner: smartWalletOwner.publicKey,
      })
      .signers([baseWallet.payer, smartWalletOwner])
      .rpc();

    return {
      sessionKey,
      userSmartWallet,
      session: sessionKeys.deriveSessionPda(
        program.programId,
        userSmartWallet,
        sessionKey.publicKey
      ),
    };
  }

  async function executeExpectingError(
    fixture: Awaited<ReturnType<typeof createApprovedSession>>,
    keys: AccountMeta[]
  ) {
    const targetInstruction = new TransactionInstruction({
      programId: TARGET_PROGRAM_PLACEHOLDER,
      keys,
      data: DUMMY_ANCHOR_DISCRIMINATOR,
    });

    return sendExpectError(
      executeWithSession(program, {
        sessionExecutor: baseWallet.publicKey,
        sessionKey: fixture.sessionKey.publicKey,
        userSmartWallet: fixture.userSmartWallet,
        targetInstruction,
      })
        .signers([baseWallet.payer, fixture.sessionKey])
        .rpc()
    );
  }

  it("Rejects the session key in remaining accounts", async () => {
    const fixture = await createApprovedSession();

    const error = await executeExpectingError(fixture, [
      {
        pubkey: fixture.sessionKey.publicKey,
        isSigner: false,
        isWritable: false,
      },
      { pubkey: fixture.userSmartWallet, isSigner: false, isWritable: false },
    ]);

    assert.ok(error.includes("RemainingAccountsContainsSessionKey"), error);
  });

  it("Rejects another program-owned account", async () => {
    const fixture = await createApprovedSession();

    const error = await executeExpectingError(fixture, [
      { pubkey: fixture.session, isSigner: false, isWritable: false },
      { pubkey: fixture.userSmartWallet, isSigner: false, isWritable: false },
    ]);

    assert.ok(
      error.includes("RemainingAccountsContainsProgramOwnedAccount"),
      error
    );
  });

  it("Rejects the smart wallet as writable", async () => {
    const fixture = await createApprovedSession();

    const error = await executeExpectingError(fixture, [
      { pubkey: fixture.userSmartWallet, isSigner: false, isWritable: true },
    ]);

    assert.ok(error.includes("UserSmartWalletAccountIsWritable"), error);
  });

  it("Rejects a missing smart wallet", async () => {
    const fixture = await createApprovedSession();

    const error = await executeExpectingError(fixture, []);

    assert.ok(error.includes("UserSmartWalletNotFound"), error);
  });

  it("Rejects a duplicate smart wallet", async () => {
    const fixture = await createApprovedSession();

    const error = await executeExpectingError(fixture, [
      { pubkey: fixture.userSmartWallet, isSigner: false, isWritable: false },
      { pubkey: fixture.userSmartWallet, isSigner: false, isWritable: false },
    ]);

    assert.ok(error.includes("MultipleUserSmartWalletAccounts"), error);
  });
});

File: programs/r3-session-keys/tests/08-execute_validations_checks.rs
mod common;

use {
    anchor_lang::{
        prelude::{AccountMeta, Pubkey},
        solana_program::instruction::Instruction,
    },
    common::{
        client::R3SessionKeysClient, send_tx_expect_error, send_tx_expect_ok,
        timestamp_from_future, Env, DUMMY_ANCHOR_DISCRIMINATOR, TARGET_PROGRAM_PLACEHOLDER,
    },
    solana_keypair::Keypair,
    solana_signer::Signer,
};

fn create_approved_session(
    env: &mut Env,
    client: &R3SessionKeysClient,
) -> (Keypair, Pubkey, Pubkey) {
    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();
    let (create_wallet, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, create_wallet, &[&env.admin]);

    let (create_session, session, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        TARGET_PROGRAM_PLACEHOLDER,
        timestamp_from_future(),
        DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
        DUMMY_ANCHOR_DISCRIMINATOR.len() as u8,
    );
    send_tx_expect_ok(&mut env.svm, create_session, &[&env.admin]);

    let approve = client.approve_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(&mut env.svm, approve, &[&env.admin, &smart_wallet_owner]);

    (session_key, user_smart_wallet, session)
}

fn execute_with_remaining_accounts(
    env: &mut Env,
    client: &R3SessionKeysClient,
    session_key: &Keypair,
    user_smart_wallet: Pubkey,
    accounts: Vec<AccountMeta>,
) -> String {
    let target_ix = Instruction {
        program_id: TARGET_PROGRAM_PLACEHOLDER,
        accounts,
        data: DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
    };
    let execute = client.execute_with_session(
        env.admin.pubkey(),
        session_key.pubkey(),
        user_smart_wallet,
        target_ix,
    );
    send_tx_expect_error(&mut env.svm, execute, &[&env.admin, session_key])
}

#[test]
fn test_execute_rejects_session_key_in_remaining_accounts() {
    let client = R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    let (session_key, user_smart_wallet, _) = create_approved_session(&mut env, &client);

    let error = execute_with_remaining_accounts(
        &mut env,
        &client,
        &session_key,
        user_smart_wallet,
        vec![
            AccountMeta::new_readonly(session_key.pubkey(), false),
            AccountMeta::new_readonly(user_smart_wallet, false),
        ],
    );

    assert!(
        error.contains("Error Code: RemainingAccountsContainsSessionKey"),
        "{error}"
    );
}

#[test]
fn test_execute_rejects_other_program_owned_account() {
    let client = R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    let (session_key, user_smart_wallet, session) = create_approved_session(&mut env, &client);

    let error = execute_with_remaining_accounts(
        &mut env,
        &client,
        &session_key,
        user_smart_wallet,
        vec![
            AccountMeta::new_readonly(session, false),
            AccountMeta::new_readonly(user_smart_wallet, false),
        ],
    );

    assert!(
        error.contains("Error Code: RemainingAccountsContainsProgramOwnedAccount"),
        "{error}"
    );
}

#[test]
fn test_execute_rejects_writable_smart_wallet() {
    let client = R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    let (session_key, user_smart_wallet, _) = create_approved_session(&mut env, &client);

    let error = execute_with_remaining_accounts(
        &mut env,
        &client,
        &session_key,
        user_smart_wallet,
        vec![AccountMeta::new(user_smart_wallet, false)],
    );

    assert!(
        error.contains("Error Code: UserSmartWalletAccountIsWritable"),
        "{error}"
    );
}

#[test]
fn test_execute_rejects_missing_smart_wallet() {
    let client = R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    let (session_key, user_smart_wallet, _) = create_approved_session(&mut env, &client);

    let error =
        execute_with_remaining_accounts(&mut env, &client, &session_key, user_smart_wallet, vec![]);

    assert!(
        error.contains("Error Code: UserSmartWalletNotFound"),
        "{error}"
    );
}

#[test]
fn test_execute_rejects_duplicate_smart_wallet() {
    let client = R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    let (session_key, user_smart_wallet, _) = create_approved_session(&mut env, &client);

    let error = execute_with_remaining_accounts(
        &mut env,
        &client,
        &session_key,
        user_smart_wallet,
        vec![
            AccountMeta::new_readonly(user_smart_wallet, false),
            AccountMeta::new_readonly(user_smart_wallet, false),
        ],
    );

    assert!(
        error.contains("Error Code: MultipleUserSmartWalletAccounts"),
        "{error}"
    );
}
